staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc_test.go

```diff
@@ -3473,6 +3499,213 @@ func TestUnmarshalClaim(t *testing.T) {
+"username": "jane",
+"exp": %d
+}`, ts.URL, valid.Unix())
+
+	jws, err := signer.Sign([]byte(claims))
+	if err != nil {
+		t.Fatalf("sign claims: %v", err)
+	}
+	token, err := jws.CompactSerialize()
+	if err != nil {
+		t.Fatalf("serialize token: %v", err)
+	}
+
+	ia, ok := a.(*instrumentedAuthenticator)
+	if !ok {
+		t.Fatalf("expected authenticator to be instrumented")
+	}
+	authenticator, ok := ia.delegate.(*jwtAuthenticator)
+	if !ok {
+		t.Fatalf("expected delegate to be Authenticator")
+	}
+	// wait for the authenticator to be initialized
+	err = wait.PollUntilContextCancel(ctx, time.Millisecond, true, func(context.Context) (bool, error) {
+		if v, _ := authenticator.idTokenVerifier(); v == nil {
+			return false, nil
+		}
+		return true, nil
+	})
+	if err != nil {
+		t.Fatalf("failed to initialize the authenticator: %v", err)
+	}
+
+	// check the provider status metrics
+	metricFamilies, err := legacyregistry.DefaultGatherer.Gather()
+	if err != nil {
+		t.Fatalf("failed to gather metrics: %v", err)
+	}
+
+	jwtIssuerHash := getHash(ts.URL)
+	for _, family := range metricFamilies {
+		if family.GetName() != "apiserver_authentication_jwt_authenticator_provider_status_timestamp_seconds" {
+			continue
+		}
+
+		if !testutil.LabelsMatch(family.Metric[0], map[string]string{
+			"apiserver_id_hash": testAPIServerIDHash,
+			"jwt_issuer_hash":   jwtIssuerHash,
+			"status":            "initializing",
+		}) {
+			t.Fatalf("unexpected metric: %v", family.Metric[0])
+		}
+
+		if !testutil.LabelsMatch(family.Metric[1], map[string]string{
+			"apiserver_id_hash": testAPIServerIDHash,
+			"jwt_issuer_hash":   jwtIssuerHash,
+			"status":            "ready",
+		}) {
+			t.Fatalf("unexpected metric: %v", family.Metric[0])
+		}
+	}
+
+	// 1. Remote JWKS fetch success the first time we fetch the keys.
+	if _, _, err = a.AuthenticateToken(ctx, token); err != nil {
+		t.Fatalf("authenticate token: %v", err)
+	}
+	checkJWKSMetrics(t, jwtIssuerHash, 5.482153416488015e+18)
+
+	// 2. Rotate the keys and update the server.
+	signingKey = loadRSAPrivKey(t, "testdata/rsa_2.pem", jose.RS256)
+	pubKeys = []*jose.JSONWebKey{
+		loadRSAKey(t, "testdata/rsa_1.pem", jose.RS256),
+		loadRSAKey(t, "testdata/rsa_2.pem", jose.RS256),
+	}
+	signer, err = jose.NewSigner(jose.SigningKey{
+		Algorithm: jose.SignatureAlgorithm(signingKey.Algorithm),
+		Key:       signingKey,
+	}, nil)
+	if err != nil {
+		t.Fatalf("initialize signer: %v", err)
+	}
+	ts.setKeys(toKeySet(pubKeys))
+	// Sign and serialize the claims in a JWT with the new key.
+	// This should trigger a new JWKS fetch as kid not found in the cache.
+	jws, err = signer.Sign([]byte(claims))
+	if err != nil {
+		t.Fatalf("sign claims: %v", err)
+	}
+	token, err = jws.CompactSerialize()
+	if err != nil {
+		t.Fatalf("serialize token: %v", err)
+	}
+	if _, _, err = a.AuthenticateToken(ctx, token); err != nil {
+		t.Fatalf("authenticate token: %v", err)
+	}
+	checkJWKSMetrics(t, jwtIssuerHash, 1.598836820000112e+19)
```

Comment:
Add tests for the failure case, and a test case with two JWT issuers so that we can assert that one metric series changing does not impact another.

Reply:
Added tests for failure: https://github.com/kubernetes/kubernetes/pull/123642/files#diff-9dfc52cf08b2714b4e54cfc514504fbf637c9182648855e7bc7a3f210d6ea63dR3611. Currently working on adding unit tests for multiple JWT issuers + metrics.

Reply:
Add metrics check to the existing multiple jwt authenticators integration test